### PR TITLE
Install middleware in ObjectAdapter (Swift)

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR improves ObjectAdapter in Swift to allow users to install middleware and construct the dispatch pipeline for the ObjectAdapter.

This is the Swift counterpart of #2277.

### CODE REVIEW

File: swift/test/Ice/middleware/AllTests.swift [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    let output = helper.getWriter()
+
+    try testMiddlewareExecutionOrder(communicator, output)
+
+    // Verifies the middleware execute in installation order.
+    func testMiddlewareExecutionOrder(_ communicator: Communicator, _ output: TextWriter) throws {
+        output.write("testing middleware execution order... ")
+
+        // Arrange
+        let oa = try communicator.createObjectAdapterWithEndpoints(name: "MyOA", endpoints: "tcp -h 127.0.0.1 -p 0")
```
Comment (bernardnormier) [inline swift/test/Ice/middleware/AllTests.swift:22]: It looks like colloc doesn't work with endpoint-less OAs in Swift.

File: swift/src/Ice/ObjectAdapter.swift [hunk 2]
```diff
@@ -56,6 +60,14 @@ public protocol ObjectAdapter: AnyObject {
     /// the same name.
     func destroy()
 
+    /// Install a middleware in this object adapter.
+    /// - Parameter middleware: The middleware to install.
+    /// - Returns: This object adapter.
+    /// - Throws: `Ice.InitializationException` if the object adapter's dispatch pipeline has already been created. This
+    /// creation typically occurs the first time the object adapter dispatches an incoming request.
```
Comment (bernardnormier) [inline swift/src/Ice/ObjectAdapter.swift:67]: Not sure if we need to document that use is not thread-safe. I.e. you shouldn't call use concurrently on the same OA.

We don't mention this "limitation" in IceRPC: 
https://github.com/icerpc/icerpc-csharp/blob/0153bc8323107da94d1fc9a360f59d737647b480/src/IceRpc/Router.cs#L117

File: swift/src/Ice/ObjectAdapterI.swift [hunk 3]
```diff
@@ -3,14 +3,30 @@
 import IceImpl
 
 class ObjectAdapterI: LocalObject<ICEObjectAdapter>, ObjectAdapter, ICEDispatchAdapter, Hashable {
-    private let communicator: Communicator
     let servantManager: ServantManager
 
+    lazy var dispatchPipeline: Dispatcher = {
+        dispatchPipelineInitialized = true
+
+        // Since this computation is idempotent, we don't need to make it thread-safe. And we don't worry about the
```
Comment (bernardnormier) [inline swift/src/Ice/ObjectAdapterI.swift:11]: This is bogus. The middleware in the dispatch pipeline can maintain state and we don't want to ever create & return two distinct pipelines.

Comment (bernardnormier) [reply swift/src/Ice/ObjectAdapterI.swift]: Fixed.

**Review by pepone — APPROVED** [review]

**Review by externl — APPROVED** [review]

**Pull request merged**